woocommerce/woocommerce-gateway-stripe
# Add a reference to the transpiled JS file in POT file reference comments

Fixes #2223

Currently the translations strings from javascript files are not being included in the POT file.

The solution is to use `wp-cli i18n command` to extract the translation strings from both PHP and JS files and generate the POT file, and then it adds additional file references to the generated (webpack transpiled) files for each string __**__.

Example POT file:
```
#: build/index.js:1
#: build/settings.js:1
#: client/additional-methods-setup/methods-selector/add-payment-methods-task.js:391
#: client/connect-account-page/modal/index.js:82
#: client/multi-currency-setup/tasks/store-settings-task/index.js:147
#: client/payment-methods/activation-modal.js:47
msgid "Continue"
msgstr ""
```

__**__: This last step is necessary because language-packs group all strings from a same `.js` filename into different JSON files (with the hash of the file reference comment appended to the JSON filename). To also generate a JSON file for the transpiled files (those enqueued by the plugin) we need to add the file references to the generated (webpack transpiled) files.

More context on the original WC Payments issue in the slack thread: p1638277640394300-slack-CGGCLBN58

#### Changes proposed in this Pull Request

This PR adds the same POT combination process (renamed to `update-pot-file-references` as we only have 1 pot file here) used in WC Payments to add a comment reference with the generated JS asset file (in the `build` folder) for each javascript source file (in the `client` folder).

It also changes the release process to generate `hidden-source-maps` during production build, which are used to append  the file references to the generated JS files to the final translations POT file.
These source-maps are not referenced in the final JS nor are included in the release zip package.

An alternative would be to use a Webpack plugin to generate the dependencies file ([like this](https://gist.github.com/diegocurbelo/b1905ab3e1cbd5c5549d9f34eca23098)), but using this plugin does not improve the build time over using hidden-source-maps:
- Develop branch:
`npm run build  113.43s user 7.20s system 274% cpu 43.928 total`
- This branch:
`npm run build  114.46s user 6.95s system 281% cpu 43.070 total`

The times for development builds are not affected, source-maps were already being generated for development builds

#### Testing instructions

1. Checkout `develop`
2. Run `npm run build`
3. Open the generated `languages/woocommerce-gateway-stripe.pot` file
4. Verify that there are no translation strings for JS files.
5. Checkout this branch
6. Run `npm run build` again
7. Open the generated `languages/woocommerce-gateway.pot` file
8. Verify that now there are translations for JS files, an also verify that they include a reference to the generated file:
```
#: build/upe_settings.js:1
#: client/settings/payment-methods/index.js:27
msgid "Payments accepted on checkout"
msgstr ""
```

#### Additional Context
- Original WC Payments issue in the slack thread: p1638277640394300-slack-CGGCLBN58 
- [babel-plugin-makepot: Don't bundle .pot files #13535](https://github.com/WordPress/gutenberg/issues/13535)
- [Update wporg-gp-customizations to generate translation files for each JS file](https://meta.trac.wordpress.org/ticket/3876)
